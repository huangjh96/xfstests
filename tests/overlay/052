#! /bin/bash
# FS QA Test No. 052
#
# Test encode/decode overlay file handles with renames
#
# - Check decode/read of file handles after rename of parent
# - Check decode/read of file handles after rename of grandparent
# - Check decode/read of file handles after move to new parent
# - Check encode/decode/read of file handles in non-upper overlay
#
# This test requires and enables overlayfs NFS export support and merge
# dir rename support (redirect_dir).
# NFS export support depends on and enables overlayfs index feature.
#
#-----------------------------------------------------------------------
# Copyright (C) 2018 CTERA Networks. All Rights Reserved.
# Author: Amir Goldstein <amir73il@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#-----------------------------------------------------------------------
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter

# real QA test starts here

_supported_fs overlay
_supported_os Linux
_require_scratch
_require_test_program "open_by_handle"
_require_scratch_feature nfs_export
_require_scratch_feature redirect_dir

# All overlay dirs are on scratch partition
lower=$OVL_BASE_SCRATCH_MNT/$OVL_LOWER
middle=$OVL_BASE_SCRATCH_MNT/$OVL_LOWER.2
upper=$OVL_BASE_SCRATCH_MNT/$OVL_UPPER
work=$OVL_BASE_SCRATCH_MNT/$OVL_WORK

NUMFILES=1

# Create test dir and empty test files
create_test_files()
{
	local dir=$1
	local opt=$2

	mkdir -p $dir
	$here/src/open_by_handle -cp $opt $dir $NUMFILES
}

# Test encode/decode file handles on overlay mount
test_file_handles()
{
	local dir=$1
	shift

	echo test_file_handles $dir $* | _filter_scratch | \
				sed -e "s,$tmp\.,,g"
	$here/src/open_by_handle $* $dir $NUMFILES
}

# Re-create lower/middle/upper/work dirs
create_dirs()
{
	_scratch_mkfs
	mkdir -p $middle
}

# Mount an overlay on $SCRATCH_MNT with all layers on scratch partition
mount_dirs()
{
	_scratch_mount -o nfs_export=on,redirect_dir=on
}

# Unmount the overlay without unmounting base fs
unmount_dirs()
{
	$UMOUNT_PROG $SCRATCH_MNT
}

# Check non-stale file handles of lower/upper moved files
create_dirs
create_test_files $upper/uppertestdir -w
create_test_files $lower/lowertestdir -w
mkdir -p $lower/lowertestdir.lo $lower/lowertestdir.up $upper/uppertestdir.lo $upper/uppertestdir.up
mount_dirs
# Check encode/decode/read of lower/upper file handles after move to new upper testdir
test_file_handles $SCRATCH_MNT/uppertestdir -o $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT/lowertestdir -o $tmp.lower_file_handles
mv $SCRATCH_MNT/uppertestdir/* $SCRATCH_MNT/uppertestdir.up/
mv $SCRATCH_MNT/lowertestdir/* $SCRATCH_MNT/uppertestdir.lo/
# Check open and read from stored file handles
test_file_handles $SCRATCH_MNT -r -i $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT -r -i $tmp.lower_file_handles
# Check encode/decode/read of lower/upper file handles after move to new merge testdir
test_file_handles $SCRATCH_MNT/uppertestdir.up -o $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT/uppertestdir.lo -o $tmp.lower_file_handles
mv $SCRATCH_MNT/uppertestdir.up/* $SCRATCH_MNT/lowertestdir.up/
mv $SCRATCH_MNT/uppertestdir.lo/* $SCRATCH_MNT/lowertestdir.lo/
# Check open and read from stored file handles
test_file_handles $SCRATCH_MNT -r -i $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT -r -i $tmp.lower_file_handles
unmount_dirs

# Check non-stale file handles of lower/upper renamed dirs
create_dirs
create_test_files $upper/uppertestdir -w
create_test_files $lower/lowertestdir -w
create_test_files $upper/uppertestdir/subdir -w
create_test_files $lower/lowertestdir/subdir -w
mount_dirs
# Check encode/decode/read of lower/upper file handles after rename of testdir
test_file_handles $SCRATCH_MNT/uppertestdir -p -o $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT/lowertestdir -p -o $tmp.lower_file_handles
# Check encode/decode/read of lower/upper file handles after rename of testdir's parent
test_file_handles $SCRATCH_MNT/uppertestdir/subdir -p -o $tmp.upper_subdir_file_handles
test_file_handles $SCRATCH_MNT/lowertestdir/subdir -p -o $tmp.lower_subdir_file_handles
# Rename pure upper dir
mv $SCRATCH_MNT/uppertestdir $SCRATCH_MNT/uppertestdir.new/
# Copy up lower dir, index and rename
mv $SCRATCH_MNT/lowertestdir $SCRATCH_MNT/lowertestdir.new/
# Check open, read and readdir from stored file handles
# (testdir argument is the mount point and NOT the dir
#  we are trying to open by stored file handle)
test_file_handles $SCRATCH_MNT -rp -i $tmp.upper_file_handles
test_file_handles $SCRATCH_MNT -rp -i $tmp.lower_file_handles
test_file_handles $SCRATCH_MNT -rp -i $tmp.upper_subdir_file_handles
test_file_handles $SCRATCH_MNT -rp -i $tmp.lower_subdir_file_handles
# Retry decoding lower subdir file handle when indexed testdir is in dcache
# (providing renamed testdir argument pins the indexed testdir to dcache)
test_file_handles $SCRATCH_MNT/lowertestdir.new -rp -i $tmp.lower_subdir_file_handles
unmount_dirs

# Check encode/decode/read of lower file handles on lower layers only r/o overlay.
# For non-upper overlay mount, nfs_export requires disabling redirect_dir.
$MOUNT_PROG -t overlay $OVL_BASE_SCRATCH_MNT $SCRATCH_MNT \
			-o ro,redirect_dir=nofollow,lowerdir=$middle:$lower
test_file_handles $SCRATCH_MNT/lowertestdir -rp
test_file_handles $SCRATCH_MNT/lowertestdir/subdir -rp
unmount_dirs

# Check encode/decode/read of lower file handles on lower layers only r/o overlay
# with non-indexed redirects from top lower layer to bottom lower layer.
# Overlay lookup cannot follow the redirect from $upper/lowertestdir.new to
# $lower/lowertestdir. Instead, we mount an overlay subtree rooted at these
# directories.
$MOUNT_PROG -t overlay $OVL_BASE_SCRATCH_MNT $SCRATCH_MNT \
		-o ro,redirect_dir=nofollow,lowerdir=$upper/lowertestdir.new:$lower/lowertestdir
test_file_handles $SCRATCH_MNT -r
test_file_handles $SCRATCH_MNT/subdir -rp
unmount_dirs

status=0
exit
