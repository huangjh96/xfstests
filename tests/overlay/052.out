QA output created by 052
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -p
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -wrap
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -p
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -wrap
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -a
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -r
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -a
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -r
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -dk
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir.rw -rwdk
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -dk
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir.rw -rwdk
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -dp
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -dp
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -wlr
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -ur
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -wlr
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -ur
test_file_handles SCRATCH_DEV/OVL_LOWER/lowertestdir -l
test_file_handles SCRATCH_DEV/OVL_UPPER/uppertestdir -l
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -wur
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -wur
test_file_handles SCRATCH_DEV/OVL_LOWER/lowertestdir -l
test_file_handles SCRATCH_DEV/OVL_UPPER/uppertestdir -l
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -d
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -d
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir -wmr
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir -wmr
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir.new -rwp
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir.new -rwp
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir.up -rp
test_file_handles SCRATCH_DEV/tmp/mnt/uppertestdir.lo -rp
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir.up -rp
test_file_handles SCRATCH_DEV/tmp/mnt/lowertestdir.lo -rp
